### ISSUE TITLE
About the inference speed

### ISSUE THREAD

**Issue by ashawkey** [kwea123/ngp_pl#5]
Congratulations on the great work!
I just have a small issue on the FPS of torch-ngp. The reported 7.8 actually includes the cost of saving rgb and depth to disk, and I estimated it to be ~20 on lego from the GUI on a V100.
Since the ray marching CUDA implementation is similar, I really wonder about the speed up, and would be grateful if you could share any ideas on how to improve it!

**Comment by kwea123** [comment]
Hi, indeed, the 7.8 includes saving so not accurate.

After commenting out these lines, I ran again with `--test` on my 2080Ti
https://github.com/ashawkey/torch-ngp/blob/main/nerf/utils.py#L848-L864

And this is the result:
![image](https://user-images.githubusercontent.com/11364490/177082300-a2620bcb-d68b-4629-ab4f-8f886dfb6a00.png)

So in average it's 200/11=18.18FPS, which is close to your number.

As for why the speed is different, during the implementation I noticed some improvements:
1.  (I'm not so familiar with cuda so this might be wrong) I pass in `PackedTensor` directly in the kernel, so I don't need to do the `locate` like [here](https://github.com/ashawkey/torch-ngp/blob/main/raymarching/src/raymarching.cu#L890-L893), this at least improves readability, and probably some speed? I'm not sure. And [here](https://github.com/ashawkey/torch-ngp/blob/main/raymarching/src/raymarching.cu#L877-L881) I use inplace addition instead of passing by additional variables.
2. I keep track of [alive indices](https://github.com/kwea123/ngp_pl/blob/master/models/rendering.py#L63) instead of making additional buffers and use atomicAdd to track [here](https://github.com/ashawkey/torch-ngp/blob/main/raymarching/src/raymarching.cu#L936-L938)
3. Probably the most effective? When marching rays in testing, we allocate a small number each time for each ray and march the rays only that little ([the variable step in your code](https://github.com/ashawkey/torch-ngp/blob/main/nerf/renderer.py#L338)). However not all rays need that many samples. Take step=1 (initial step) for example, we march 1 sample further for each ray, but many rays don't even hit anything, so we can exclude those samples that hits nothing from network evaluation like
https://github.com/kwea123/ngp_pl/blob/ebd45392c2b8dd58579ee317102fe731d259568b/models/rendering.py#L80-L87
In [your code](https://github.com/ashawkey/torch-ngp/blob/main/nerf/renderer.py#L357-L359), you evaluate every samples, among which many are useless. Take the first step as example, there are 800x800x1 samples, but if we mask out the samples that hit nothing, in my experiment only ~200k samples need to be evaluated, and that leads to speedup.

I can only think of these small things that are different, but unsure if they are really the difference.

**Comment by ashawkey** [comment]
Thanks for the detailed explanation! I'll experiment on them later.

**Comment by kwea123** [comment]
A notebook test.ipynb and a pretrained model is uploaded, you can use them to evaluate and measure the fps.

**Comment by ashawkey** [comment]
Yes I'm doing so! It seems the `T_threshold` also matters.

* `1e-2`: 33.59
* `1e-4`: 28.41
* `1e-2` w/o masking (point 3): 33.30
* `1e-4` w/o masking (point 3): 28.10

(The results are from a TITAN RTX)

by `w/o masking` I modified the code to:
```python
        sigmas, rgbs = model(xyzs, dirs)
        sigmas = sigmas.contiguous().float()
        rgbs = rgbs.contiguous().float()
```

Some quick experiments on torch-ngp also show point 2 and 3 may be not very effective. 
I'm not very sure about the difference between `PackedTensor` and `data_ptr` for point 1, but the indexing overhead should be small.

Also, this notebook only measures rendering time, but the data loading strategy also matters.
ngp_pl preprocess images into rays, while torch_ngp generates rays on-the-fly (a time-memory trade-off).

Finally, thanks for the information and help! It seems still difficult to match native instant-ngp's speed while keeping enough flexibility.

P.S. a little name inconsistency in [released ckpts](https://github.com/kwea123/ngp_pl/releases/tag/v1.0), it should be `epoch=19_slim.ckpt` to match the notebook.

**Comment by kwea123** [comment]
I just noticed a big bottleneck in data loading, I fixed it and now it trains 1.5x faster. I will re-benchmark everything

**Comment by SSground** [comment]
Whether to consider using volrender to improve inference speed？
